File: 30days-of-code/22_binarytree.py
class Node:
    def __init__(self, value):
        self.right = self.left = None
        self.data = value

class Solution:
    def insert(self, root, value):
        if root is None:
            return Node(value)

        if value <= root.data:
            cur = self.insert(root.left, value)
            root.left = cur
        else:
            cur = self.insert(root.right, value)
            root.right = cur
        return root


    def get_height(self, root_node):
        def longest_path(node, height):
            left_height = right_height = 0
            if node.left is not None:
                left_height = longest_path(node.left, height)
                left_height += 1
            if node.right is not None:
                right_height = longest_path(node.right, height)
                right_height += 1
            cur_height = max(left_height, right_height) + height
            return cur_height

        if root_node is None:
            return 0

        return longest_path(root_node, 0)


    def level_order(self, root):
        if root is None: return []

        levels = []
        next_level = []
        current_level = [root]
        while current_level:
            levels.append(current_level)
            next_level = []
            for node in current_level:
                if node.left is not None: next_level.append(node.left)
                if node.right is not None: next_level.append(node.right)
            current_level = next_level

        return levels


def read_input():
    n_lines = int(input())
    values = []
    for _ in range(n_lines):
        values.append(int(input()))
    return values


if __name__ == "__main__":
    # this is global namespace !!
    data = read_input()

    my_tree = Solution()
    tree_root = None
    for elem in data:
        tree_root = my_tree.insert(tree_root, elem)

    tree_height = my_tree.get_height(tree_root)
    print(tree_height)

    tree_levels = my_tree.level_order(tree_root)
    print(' '.join(map(str, (node.data for alevel in tree_levels for node in alevel))))

File: 30days-of-code/22_binarytree.cc
#include <iostream>
#include <vector>

using namespace std;    


class Node {
    public:
    const int data;
    Node* left;
    Node* right;

    public:
    explicit Node(const int d) : data{d}, left{nullptr}, right{nullptr}
    {}
};


class BinaryTree {
    public:
    Node* insert(Node* root, const int data) {
        if (not root) {
            return new Node(data);
        }

        Node* cur{};
        if (data <= root->data) {
            cur = insert(root->left, data);
            root->left = cur;
        }
        else {
            cur = insert(root->right, data);
            root->right = cur;
       }

       return root;
    }

    int getHeight(const Node* const root) const {
        if (not root) {
            return 0;
        }
        return walk(root, 0);
    }

    vector<vector<Node*>> listBreadthFirst(Node* root) const {
        if (not root) {
            return vector<vector<Node*>>{};
        }

        vector<vector<Node*>> levels{};
        vector<Node*> nextLevel{};
        vector<Node*> currentLevel{root};

        while (not currentLevel.empty()) {
            levels.push_back(currentLevel);
            vector<Node*>{}.swap(nextLevel);
            for (auto node : currentLevel) {
                if (node->left) {
                    nextLevel.push_back(node->left);
                }
                if (node->right) {
                    nextLevel.push_back(node->right);
                }
            }
            currentLevel = nextLevel;
        }

        return levels;
    }

    private:
    int walk(const Node* const node, const int height) const {
        int leftHeight{0}, rightHeight{0};
        if (node->left) {
            leftHeight = walk(node->left, height);
            ++leftHeight;
        }
        if (node->right) {
            rightHeight = walk(node->right, height);
            ++rightHeight;
        }
        return max(leftHeight, rightHeight) + height;
    }

};


vector<int> readMany() {
    int t{};
    cin >> t;

    int value{};
    vector<int> values{};
    while(t-- > 0){
        cin >> value;
        values.push_back(value);
    }
    return values;
}


int main() {
    vector<int> items = readMany();

    BinaryTree myTree{};
    Node* root{nullptr};
    for (auto item : items) {
        root = myTree.insert(root, item);
    }

    int height = myTree.getHeight(root);
    cout << height << endl;

    vector<vector<Node*>> treeLevels = myTree.listBreadthFirst(root);
    for (auto level : treeLevels) {
        for (auto node : level) {
            cout << node->data << " ";
        }
    }
    cout << endl;

    return 0;
}
